Revamp (again) API defaults and validations (#4672)
aws/eks-anywhere: designs/validations.md

* `test/framework`
	* `ClusterValidator` is yet another validator runner. This is one centered in validating the state of a cluster against a given cluster specification by using a k8s client to inspect the cluster API.
* Special note is due for commands in `cmd`, since these are the entities with more knowledge and they are a good example of the fragmentation addressed in this doc. Currently, they handle different types of validations and at different stages:
	* Independently run validations, like:
		* `FileExists` for the cluster config yaml.
		* Parsing and Cluster config API validation (`GetAndValidateClusterConfig`) 
		* Tinkerbell CLI flags.
		* Registry mirror auth.
		* Validations to `cluster.Config` with `cluster.Validate`. Defaults are applied by just calling `cluster.NewSpecFromClusterConfig`.
	* "Grouped" validations in `commonValidations`
		* Docker daemon config
		* File exists, parsing and Cluster config API validation (`GetAndValidateClusterConfig`). This is duplicated.
		* Kubeconfig file and format validation.
	* Delegated validations to the workflows through `interfaces.Validator`.

## Goals

> * Define a standard for types of validations/defaults and where and **where not** to put them.

Reviewer: We might want to replace the `/` in some of these goals to make clear we're addressing 2 independent concepts, not a pseudonym as `/` can represents.

E.g.

```
* Define a standard for types of validation and defaults.
* Define where validations and defaults should be defined within the code base.
```

### Implementation (*where*)
Static data validations/defaults that only require one single API object **should** live in the `Validate` and `SetDefaults` api struct methods, thus in the `pkg/api/v1**` package. (*Note:* I'm more than open to relax this constrain if there is a good reason to, but given this is one of the only things that seem to resemble some kind consistency with regards to this whole topic, it seems worth to maintain it).

Immutability validations have been treated differently (for some unknown reason) in the CLI than in webhooks. These checks **should** be implemented in the `api` package and **should not** be duplicated in two places. If the same checks are needed in both webhook and CLI, implement as standalone functions and reuse in both places. In the CLI, just read the object with a client and pass both old and new like the webhook does.

The rest of validations and defaults (static data for a group of API structs and the context aware ones) **can** live anywhere, as long the dependency directions make sense, with a few constrains:
* Context aware validation **cannot** live in `pkg/api/v1**`. This is to keep this package dependencies at a minimum.

> * **Avoid** packages like `validations`. Don't group validations for what they are, group them by what they validate (their context). For example, for docker validations, don't use `validations.ValidateDockerDesktopVersion`, think about `docker.ValidateDesktopVersion` in `pkg/docker`.

Reviewer: I think we expect the `executables` package to be broken up (we'll probably have a separate discussion on how to organize the broader code base but this proposal gives hints on the general direction). We may, eventually, see `type Client interface` and `type Executable struct` both in `/pkg/docker`.

The `executables` package is representative of a layered architecture. E.g. In MVC, models, views, and controllers each have their own respective packages. That's generally not the norm for organizing Go programs.

* **Avoid** the `cmd` package. These tend to end up being untested and lack structure.

Webhooks that only require static data validations/defaults **should** live in the `pkg/api/v1**` (where all present webhooks live today). If they have any other dependency the **cannot** live in the `api` package. (*Note*: we could possibly define a `webhooks` folder elsewhere if we think that could avoid problems in the future, leaving this up for discussion)

### Scope (*what*)
#### Defaults
* Defaults should not **modify** data provided by the user, they can only **add** information. Any transformation needed should be moved to runtime operations and never get reflected back on the API objects.
* They should have not secondary effects on anything else except the provided API object/s, including the internal state of the entity running the validation.
* API objects should only be modified by the defaulting logic. This guarantees they are invariants during the CLI execution, allowing for workflow reentry/re-execution (think command fails and on the next execution we continue from where it stopped).

#### Validations

> * They should have not secondary effects on **anything**, including the internal state of the entity running the validation.

Reviewer: ```suggestion
* They must not have secondary effects on such as modifying the object under validation.
```

* API struct `Validate` methods should return a [`field.ErrorList`](https://pkg.go.dev/k8s.io/apimachinery/pkg/util/validation/field#ErrorList). This facilitates using it from webhooks and ensures the API user facing error is as complete as possible.

> ### Calling (*when*)

Reviewer: Do we want to address the rare cases we have where some CLI flag parsing code wants needs to know what provider is in-use before deciding if a flag is required?

Author: Ok, if we can get an light consensus quickly on in, I'm down. here is my proposal (feel free to tweak wording):

> * In the special case where the code parsing/validating another type of CLI user input ( one that is not part of the EKS-A API, for example, command flags) requires information only available in the API (cluster config file):
>   1. Read the cluster config file
>   1. Parse the yaml content into a `cluster.Config`
>   1. Parse/validate the command flags
>   1. Construct a `cluster.Spec` with this `cluster.Config`
>   1. Continue with API defaults, validations and the rest of the command flow.
>   * Note how this requires complete decoupling between parsing the yaml content and the defaults and validations. Something that historically we haven't followed.

Reviewer: It was just a thought I had. It can totally be addressed independently.

#### Defaults
 * CLI
	* At the beginning of the command flow, in most cases before any other business logic.
	* Always before validations.
	* This doc doesn't prescribe which entity should trigger the defaults (command in `cmd` package, workflow, etc.). But they do need to be run before the API objects get injected in any other entity as a dependency (think *service* objects). Given the approach the provider's refactor project has taken to inject the `cluster.Spec` and any other non dynamic data (this is, known at the beginning of the command execution) as state in the service objects, `cmd`  seems like the obvious place to do this since this is where service objects are constructed.

> * API

Reviewer 2: Is this also before respective validations in webhook and controllers or not the same as cli?

Author: Sorry, I don't follow. The API bullet point only refers to "webhooks + controllers"

> 	* Webhook: static data defaults (only depend on the given object) and sometimes, defaults that only require kubernetes API calls (to the same cluster where the object lives). Remember that only the given object should be altered from the webhook.

Reviewer 2: By default, the webhook doesn't have the client object injected to be able to call regular kubernetes API calls. Does this call for creating a custom webhook handler or do we have an example in our code that can in fact make kubernetes API calls?

Author: We don't have an example, we haven't needed this yet (or decided not to do it).
As of how to write this kind of webhook, we just need to implement the `webhook.CustomDefaulter` interface from controller-runtime and register it with `NewWebhookManagedBy(mgr)` against our controller manager.

	* Controllers: any other defaults requiring extra information, calls to external systems or simply *too slow* for a webhook.

#### Validations
*  CLI
	* At the beginning of the command flow, in most cases before any other business logic.
	* Always after defaults.

> 	* This doc is not perspective as of which entity should trigger the validations (command in `cmd` package, workflow, etc.). The decision is left for the provider refactor team to make later.

Reviewer: ```suggestion
	* This doc does not prescribe the entity invoking validations (command in `cmd` package, workflow, etc.). The decision is left for the provider refactor team to make later.
```

* API
	* Webhook: static data validations (only depend on the given object) and sometimes, validations that only require kubernetes API calls (to the same cluster where the object lives).
		* When using a client to read from the API server, be very careful. Webhooks are supposed to be very fast, remember they are synchronous.
			* Only run *efficient* queries (avoid full scans) and always think about the number of results you expect to get (depending on the type, the difference can be of orders of magnitude).
			* If your design performance is not very obvious, benchmark it.
	* Controllers: any other validations requiring extra information, calls to external systems or simply *too slow* for a webhook.

### Composing validations
This section lays down the requirements for a framework for validations. Some requisites:
* Allow to compose and run validations implemented in different places.
* Make implementing validations as easy as possible.
* Design mainly for the CLI use-case. Optimize defaults for this scenario and possibly allow custom configuration for others.

> * We want to follow a similar idea to what `cluster.ConfigManager` and other attempts to validation framework propose: a two step process, where validations/defaults can be registered and then run:

Reviewer: ```suggestion

This approach is similar to `cluster.ConfigManager`, and some other approaches, in that it takes a two step process where validations are registered before they are run.
```

```go

> runner := validation.NewRunner()

Author: Added something, let me know

This boils down to take a context and a cluster spec as input and return some type of error as output. We could implement it with functions or an interface, but given that we want to make registering validations as easy as possible and keep the framework API surface as simple as possible, we opt out for a function:

```go
type Validation func(ctx context.Context, spec *cluster.Spec) error
```

#### Composability
We want the framework to allow, not only to register validations implemented in different places, but also validations that are composed of multiple validations. This, although it introduces some complexity, it offers a lot of flexibility to registrars to aggregate validations in different ways and it facilitates reusing existing validations (that tend to be already aggregated in one big method).

##### Aggregating errors
For validations to be registered as one handler but still be considered as different in the runner result, they should return an `errors.Aggregate`. The framework will understand this and will flatten the result. This allows for multiple levels of nesting.

##### Concurrency

> Most *preflight* validations we currently run in the CLI can be run concurrently as they are not depending on each other. Given most of the time is spent waiting for external calls (mostly HTTP), this could potentially speed up the process.

Reviewer: ```suggestion
Most *preflight* validations we currently run in the CLI can be run concurrently as they are not dependent on each other. For validations that make external calls this can be slow, concurrent invocation of validations offers a convenient speed improvement.
```

	* For debug logging (we can generalize this to anything above verbosity `0`), validations should use their own logger. This should be injected at construction time as any other dependency. This still gives control to the program to decide which logs are presented to the user and when, by passing different loggers.
	* For information more tailored to feedback for the end user (like progress in slow validations), it will be left up to the framework callers to implement this. Based how logging is currently used to communicate this progress, the information is usually presented by "grouping" validations and not at the individual validation level. So the *usecases*, where validations are grouped and given semantic value, will be responsible to define these different groups and print logs in between them as necessary. This can be easily accomplished by composing multiple runners. Given a traditional logger might not be the best way to implement a presentation layer of this kind, this approach also allows to replace the logger with something else (in the future) without having to refactor any of the validation logic.

#### Validation error meta information

> The existing `validations.ValidationResult` allows handlers to return a `Remediation` message. Given we want to just use the `error` interface to make it easy to register validations and that `Remediation` is empty for most `ValidationResult`'s , we chose to make this optional. Validations wanting to provide a possible remediation for a failure, can return an `error` that also implements the `Fixable` interface.

Reviewer 2: ```suggestion
The existing `validations.ValidationResult` allows handlers to return a `Remediation` message. Given we want to just use the `error` interface to make it easy to register validations and that `Remediation` is empty for most `ValidationResult`s , we chose to make this optional. Validations wanting to provide a possible remediation for a failure, can return an `error` that also implements the `Fixable` interface.
```

Reviewer 2: Also I am confused by this, so where is `Fixable` returned for the validator runner to process remediation and print? Or is that the job of the framework callers?

Reviewer 2: I think I see it now in the example below, but essentially we check for whether this is a Fixable interface or not with remediation?

Reviewer: > I think I see it now in the example below, but essentially we check for whether this is a Fixable interface or not with remediation?

That's the gist.

Author: The runner doesn't process errors or print them or do anything with them except flatten them if they are aggregated. It's up to the caller to decide what to do with those errors.

And exactly, we offer a util method that inspects errors to see if they implement `Fixable` or not, and if they do, return the remediation message. That is supposed to be used by the caller, in case it's interested in that kind of info.

```go

> type Fixable type {

Reviewer: Would you be open to `Remediable`?

Author: Yes, changing

	Remediation() string
}
```

#### Retrying
Existing validation runners, like `framework.ClusterValidator`, incorporate retries into the runner itself. That might seem sensible, given a lot of validations might want to rely on retries to bypass transient errors. However, not only this would complicate the API to register validations but the retry logic will always be limited by the API itself (which is only addressable by making the API more complex).

In this case we would leave this out of the framework and let each validation handler decide how to deal with retries.

#### Framework implementation
Given the framework logic is very much independent from the object being validated (`cluster.Spec`), we propose to use generics so the same framework can be used for different types. The complexity added is minimal and even in the case where the framework grows in a direction where generics don't make sense, moving back to a concrete type would be trivial. There are already places in the codebase where we envision make use of this:

```go
// Validation is the logic for a validation of a type O.
type Validation[O any] func(ctx context.Context, obj O) error
```

> In order to be able to prevent validations from modifying `obj` and detecting when if that happens, we add a constrain to `O` that allows us to `DeepCopy` the object so we can compare after running all validations:

Reviewer 2: In what situation do we need to compare the objects after running all validations? It would only be during right, and as you mentioned above the validations should not be modifying the objects.

Reviewer: > In what situation do we need to compare the objects after running all validations?

Every time the set of validations run. The goal is to make it very difficult if not impossible to abuse validations ensuring the have an appropriately narrow scope.

```go
type Validatable[O] interface {
	DeepCopy() O
}

// Validation is the logic for a validation of a type O.
type Validation[O Validatable[O]] func(ctx context.Context, obj O) error
```

### Composing defaults
#### Defaults signature
The requirements for defaults are quite similar to the ones for validations:
* Take a `context.Context`
* Based on `cluster.Spec`
* Be able to return an error if, for example, an external call fails.

In addition, and following good practices for functions that are supposed to update a given object:
* Don't rely on updating a given pointer value but take a `cluster.Spec` as input and return a `cluster.Spec` as output, with the necessary updates.

```go
type Default func(ctx context.Context, spec cluster.Spec) (cluster.Spec, error)
```

##### Problems with this approach

> Defaults should only be set in the API objects configurable by the user. For example, the `Bundles` (which compiles all eks-a dependencies) shouldn't be modified. However, all of these are part of the `cluster.Spec`, giving defaulting logic de ability to modify them. We could enforce this by separating them and providing a `cluster.Config` (all user inputable API objects) and a `Bundles` (or something else that includes the rest of the cluster specification).

Reviewer 2: ```suggestion
Defaults should only be set in the API objects configurable by the user. For example, the `Bundles` (which compiles all eks-a dependencies) shouldn't be modified. However, all of these are part of the `cluster.Spec`, giving defaulting logic the ability to modify them. We could enforce this by separating them and providing a `cluster.Config` (all user inputable API objects) and a `Bundles` (or something else that includes the rest of the cluster specification).
```

#### Framework implementation
Following the same logic as for validations, we propose the use of generics:

```go
// Default is the logic for a default for a type O. It should return a value of O
// wether it updates it or not. When there is an error, return the zero value of O
// and the error.
type Defaulter[O any] func(ctx context.Context, obj O) (O, error)
```

#### Concurrency
All defaults will be sequential given that they need to update the `cluster.Spec` and changes need to be propagated. However, defaults shouldn't rely on order. If such dependency exists between two or more defaults, then make them one single default.

> Default handlers can chose to implement concurrency but they are responsible for handling data races.

Reviewer 2: ```suggestion
Default handlers can choose to implement concurrency but they are responsible for handling data race conditions.
```
